java-checks/src/main/java/org/sonar/java/checks/SpecializedFunctionalInterfacesCheck.java

```diff
@@ -0,0 +1,239 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.ListTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeArguments;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.WildcardTree;
+
+@Rule(key = "S4276")
+public class SpecializedFunctionalInterfacesCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.VARIABLE);
+  }
+
+  private static final String REPORT_STRING = "Refactor this code to use the more specialised Functional Interface";
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.CLASS)) {
+      checkClassInterfaces(((ClassTree) tree).superInterfaces());
+    } else if (tree.is(Tree.Kind.VARIABLE)) {
+      checkVariableTypeAndInitializer((VariableTree) tree);
+    }
+  }
+
+  public void checkClassInterfaces(ListTree<TypeTree> classInterfaces) {
+    List<InterfaceTreeAndStringPairReport> reportTreeAndStringInterfaces = classInterfaces.stream().map(SpecializedFunctionalInterfacesCheck::matchFunctionalInterface)
+      .filter(Optional::isPresent)
+      .map(Optional::get)
+      .collect(Collectors.toList());
+    if (!reportTreeAndStringInterfaces.isEmpty()) {
+      List<JavaFileScannerContext.Location> flow = reportTreeAndStringInterfaces.stream()
+        .map(interf -> new JavaFileScannerContext.Location("Interface Not Specialized", interf.classInterface))
+        .collect(Collectors.toList());
+      if (reportTreeAndStringInterfaces.size() > 1) {
+        reportIssue(((ClassTree) classInterfaces.parent()).simpleName(),
+          REPORT_STRING + "s '" + reportTreeAndStringInterfaces.stream().map(x -> x.reportString).collect(Collectors.joining(", ")) + "'", flow, null);
+      } else {
+        reportIssue(((ClassTree) classInterfaces.parent()).simpleName(), REPORT_STRING + " '" + reportTreeAndStringInterfaces.get(0).reportString + "'", flow, null);
```

Comment:
you should extract a method to get the report message, it would simplify your life:`reportMessage(String ... interfaces)`

```
private static String reportMessage(String interface) {
  reportMessage(Collections.singletonList(interface));
}

private static String reportMessage(List<String> interfaces) {
    String functionalInterfaces = interfaces.stream().collect(Collectors.joining(
      "', '",
      (interfaces.size() > 1 ? "s" : "") + " '",
      "'."));
    return String.format("Refactor this code to use the more specialised Functional Interface%s", functionalInterfaces);
  }

```